### DISCUSSION TITLE
为什么LogInput::HoldOn 不用trylock?

### DISCUSSION THREAD

**Discussion by nelsontang** [alibaba/ilogtail#1912] [Help] [answered]
在采集数据和checkpoint dump频繁的情况下，发现很大概率会出现LogInput::HoldOn()死锁，死锁部分在pthread_rwlock_wrlock，这里为什么不用 pthread_rwlock_trywrlock ？

死锁是打印的部分堆栈，日志会停留在 event handle daemon pause:starts

Thread 1 (Thread 0x7fc08f9f5700 (LWP 300868)):
#0  0x00007fc08f1c62ce in pthread_rwlock_wrlock () from /lib64/libpthread.so.0
#1  0x00000000007657b0 in logtail::ReadWriteLock::lock() ()
#2  0x0000000000822b43 in logtail::LogInput::HoldOn() ()
#3  0x00000000009ac6ab in logtail::FileServer::PauseInner() ()
#4  0x00000000009ac4fb in logtail::FileServer::Pause(bool) ()
#5  0x000000000091f59c in logtail::EventDispatcher::DumpCheckPointPeriod(int) ()
#6  0x000000000072d741 in logtail::Application::Start() ()
#7  0x00000000006d5be7 in do_worker_process() ()
#8  0x00000000006d5df1 in main ()

**Comment by yyuuttaaoo** [comment]
请问是什么版本，能否详细描述一下复现步骤。这里需要停止input，所以需要用锁block线程。

**Reply by nelsontang** [reply]
ilogtail 1.8，采集一个通配目录，文件很多，在某一台机器上跑2小时左右会复现，其它机器暂时没有出现，复现前出现较多Add block event。最后卡在pthread_rwlock_wrlock，尚不明确触发条件，先问下有没有遇到过类似问题

**Reply by yyuuttaaoo** [reply, marked as answer]
1.8本身是没有问题的
main分支上这个问题在 #1825 修复了，请使用2.0分支或2.1分支最新代码，如果是自己fork的版本请pick这个patch
